### PR TITLE
Change data-type from mixed to composite and date-time to string on the fly for REST API

### LINKED ISSUE #27282: REST-API broken after update from 4.3.1 → 4.3.2 (Error: Invalid parameter(s): meta_data [rest_invalid_param])
Please provide us with the information requested in this bug report. Without these details, we won't be able to fully evaluate this issue. 
Bug reports lacking detail, or for any other reason than to report a bug, may be closed without action.

**Prerequisites (mark completed items with an [x]):**
- [ ] I have have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
After updating our store from 4.3.1 to 4.3.2 I'm unable to apply changes through the REST-API version `wc/v2` using [WooCommerce API - PHP Client](https://github.com/woocommerce/wc-api-php) v1.3.

Following payload get's transmitted to the PHP Client (replaced some real customer information with fake infos):

```
{
    "id":11089,
    "parent_id":0,
    "number":"11089",
    "order_key":"wc_order_BejTVsekPdl1b",
    "created_via":"checkout",
    "version":"3.5.7",
    "status":"processing",
    "currency":"EUR",
    "date_created":"2019-03-15T10:12:16",
    "date_created_gmt":"2019-03-15T09:12:16",
    "date_modified":"2019-03-15T10:12:21",
    "date_modified_gmt":"2019-03-15T09:12:21",
    "discount_total":"0.00",
    "discount_tax":"0.00",
    "shipping_total":"2.95",
    "shipping_tax":"0.00",
    "cart_tax":"3.98",
    "total":"27.90",
    "total_tax":"3.98",
    "prices_include_tax":true,
    "customer_id":0,
    "customer_ip_address":"127.0.0.1",
    "customer_user_agent":"Mozilla\/5.0 (Linux; Android 8.1.0; SM-J530F) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/72.0.3626.121 Mobile Safari\/537.36",
    "customer_note":"",
    "billing":{
        "first_name":"firstname",
        "last_name":"lastname",
        "company":"",
        "address_1":"address",
        "address_2":"",
        "city":"city",
        "state":"",
        "postcode":"12345",
        "country":"DE",
        "email":"firstname@example",
        "phone":"012346576",
        "title":"1",
        "vat_id":""
    },
    "shipping":{
        "first_name":"firstname",
        "last_name":"lastname",
        "company":"",
        "address_1":"address",
        "address_2":"",
        "city":"city",
        "state":"",
        "postcode":"12345",
        "country":"DE"
    },
    "payment_method":"ppec_paypal",
    "payment_method_title":"PayPal",
    "transaction_id":"sdfsd123123123",
    "date_paid":"2019-03-15T10:12:21",
    "date_paid_gmt":"2019-03-15T09:12:21",
    "date_completed":null,
    "date_completed_gmt":null,
    "cart_hash":"a035c8b3fdc20582f787067f69566744",
    "meta_data":[
        {
            "id":126548,
            "key":"_billing_title",
            "value":"1"
        },
        {
            "id":126549,
            "key":"_shipping_title",
            "value":"1"
        },
        {
            "id":126551,
            "key":"wpml_language",
            "value":"de"
        },
        {
            "id":126552,
            "key":"_paypal_status",
            "value":"completed"
        },
        {
            "id":126553,
            "key":"_woo_pp_txnData",
            "value":{
                "refundable_txns":[
                    {
                        "txnID":"123213abc",
                        "amount":"27.90",
                        "refunded_amount":0,
                        "status":"Completed"
                    }
                ],
                "txn_type":"sale"
            }
        },
        {
            "id":126559,
            "key":"_paypal_transaction_fee",
            "value":"1.04"
        },
        {
            "id":126560,
            "key":"Payer PayPal address",
            "value":"firstname@example.com"
        },
        {
            "id":126561,
            "key":"Payer first name",
            "value":"firstname"
        },
        {
            "id":126562,
            "key":"Payer last name",
            "value":"lastname"
        },
        {
            "id":126563,
            "key":"Payment type",
            "value":"instant"
        },
        {
            "id":145269,
            "key":"_direct_debit_holder",
            "value":""
        },
        {
            "id":145270,
            "key":"_direct_debit_iban",
            "value":""
        },
        {
            "id":145271,
            "key":"_direct_debit_bic",
            "value":""
        },
        {
            "id":145272,
            "key":"_direct_debit_mandate_id",
            "value":""
        },
        {
            "id":145273,
            "key":"_billing_vat_id",
            "value":""
        }
    ],
    "line_items":[
        {
            "id":4638,
            "name":"Sample product - Product",
            "product_id":610,
            "variation_id":611,
            "quantity":1,
            "tax_class":"",
            "subtotal":"20.97",
            "subtotal_tax":"3.98",
            "total":"20.97",
            "total_tax":"3.98",
            "taxes":[
                {
                    "id":35,
                    "total":"3.9843",
                    "subtotal":"3.9843"
                }
            ],
            "meta_data":[
                {
                    "id":73203,
                    "key":"pa_quality",
                    "value":"soft-toy"
                },
                {
                    "id":73204,
                    "key":"_units",
                    "value":""
                },
                {
                    "id":73205,
                    "key":"_delivery_time",
                    "value":""
                },
                {
                    "id":73206,
                    "key":"_item_desc",
                    "value":""
                },
                {
                    "id":73207,
                    "key":"_unit_price",
                    "value":""
                },
                {
                    "id":73208,
                    "key":"_unit_price_excl",
                    "value":""
                },
                {
                    "id":73209,
                    "key":"Type",
                    "value":"toy"
                },
                {
                    "id":73210,
                    "key":"Cart id",
                    "value":"31cf1ec68e1ee40e3e0a928ab3634879"
                },
                {
                    "id":73211,
                    "key":"Begin",
                    "value":""
                },
                {
                    "id":73212,
                    "key":"Middle",
                    "value":""
                },
                {
                    "id":73213,
                    "key":"End",
                    "value":""
                },
                {
                    "id":73214,
                    "key":"State",
                    "value":null
                },
                {
                    "id":73215,
                    "key":"Month",
                    "value":null
                },
                {
                    "id":73216,
                    "key":"Year",
                    "value":null
                },
                {
                    "id":73217,
                    "key":"BEIDEN SEITEN GLEICH",
                    "value":""
                },
                {
                    "id":73218,
                    "key":"Logo",
                    "value":""
                },
                {
                    "id":73219,
                    "key":"Font Family",
                    "value":null
                },
                {
                    "id":73220,
                    "key":"Font Color",
                    "value":null
                },
                {
                    "id":73221,
                    "key":"Font Texts",
                    "value":null
                },
                {
                    "id":73222,
                    "key":"Country Flag",
                    "value":null
                },
                {
                    "id":73223,
                    "key":"Size",
                    "value":"101x29"
                },
                {
                    "id":73224,
                    "key":"Back-Manufacturer",
                    "value":null
                },
                {
                    "id":73225,
                    "key":"Back-Model",
                    "value":null
                },
                {
                    "id":73226,
                    "key":"Back-Construction Year",
                    "value":""
                },
                {
                    "id":73227,
                    "key":"Back-Car Color",
                    "value":"000000"
                },
                {
                    "id":73228,
                    "key":"Back-View",
                    "value":null
                },
                {
                    "id":73229,
                    "key":"Back-Background",
                    "value":""
                },
                {
                    "id":73230,
                    "key":"Animal",
                    "value":"bear"
                },
                {
                    "id":73231,
                    "key":"Quality",
                    "value":"soft-toy"
                },
                {
                    "id":73232,
                    "key":"Image Size",
                    "value":"65x65"
                },
                {
                    "id":73233,
                    "key":"User Image",
                    "value":"https:\/\/example.com\/wp-content\/uploads\/user\/imagejpg"
                },
                {
                    "id":73234,
                    "key":"Sample Text",
                    "value":"What ever you want"
                },
                {
                    "id":85473,
                    "key":"pa_quality",
                    "value":"soft-toy"
                },
                {
                    "id":85474,
                    "key":"pa_quality",
                    "value":"soft-toy"
                }
            ],
            "sku":"",
            "price":20.97
        }
    ],
    "tax_lines":[
        {
            "id":4640,
            "rate_code":"DE-MWST. DE-1",
            "rate_id":35,
            "label":"MwSt. DE",
            "compound":false,
            "tax_total":"3.98",
            "shipping_tax_total":"0.00",
            "rate_percent":0,
            "meta_data":[

            ]
        }
    ],
    "shipping_lines":[
        {
            "id":4639,
            "method_title":"Versandkostenpauschale",
            "method_id":"flat_rate",
            "instance_id":"1",
            "total":"2.95",
            "total_tax":"0.00",
            "taxes":[

            ],
            "meta_data":[
                {
                    "id":73240,
                    "key":"Positionen",
                    "value":"Sample product &#8211; Product &times; 1"
                }
            ]
        }
    ],
    "fee_lines":[

    ],
    "coupon_lines":[

    ],
    "refunds":[

    ],
    "shipments":[

    ],
    "currency_symbol":"\u20ac",
    "parcel_delivery_opted_in":"",
    "direct_debit":{
        "holder":"",
        "iban":"",
        "bic":"",
        "mandate_id":""
    },
    "needs_confirmation":false,
    "_links":{
        "self":[
            {
                "href":"https:\/\/example.com\/wp-json\/wc\/v2\/orders\/11089"
            }
        ],
        "collection":[
            {
                "href":"https:\/\/example.com\/wp-json\/wc\/v2\/orders"
            }
        ]
    }
}
```

**Expected behavior**
Sended changes should get applied as with v4.3.1

**Actual behavior**
Error is same as described in #18810, but the transmitted `meta_data` did work with v4.3.1 and get's invalid with v4.3.2.
My guess: Update of `woocommerce/woocommerce-rest-api` from `1.0.10` to `1.0.10-pl-1` introduced a changed behavior.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
1. send payload to wocommerce v4.3.2
2. receive error:

```
Error: Invalid parameter(s): meta_data [rest_invalid_param]
``` 

**Screenshots**

**Isolating the problem (mark completed items with an [x]):**
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
We use the [WooCommerce System Status Report](https://docs.woocommerce.com/document/understanding-the-woocommerce-system-status-report/) to help us evaluate the issue. 
Without this report we won't be able to fully evaluate this issue.
<details>
```
The System Status Report is found in your WordPress admin under **WooCommerce > Status**. 
Please select “Get system report”, then “Copy for support”, and then paste it here.
```
</details>

### PR DESCRIPTION
**Pull request by vedanshujain**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

There are few changes in this PR:

1. Reverted change where we converted from mixed to built-in data types: https://github.com/woocommerce/woocommerce/commit/083e529668d5d8b8a72d829253ca3ff82430c647
2. Added on the fly conversions for date-time to string
3. Added on the fly conversions for mixed to composite data types.

### Important
When converting from mixed type, new types have to be converted into `array( 'null', 'object', 'string', 'number', 'boolean', 'integer', 'array' );` in **this order**. Otherwise, WP Core will probably cast them differently, for example, string can be cast as an array, boolean can be cast as int, and so on. This is the only order I could find which does seem to be working as expected  and is from most specific to least specific as per implementation of core's sanitization functions.

Closes #27282.

### How to test the changes in this Pull Request:

For any product, run an API request like so:
```
PUT https://w.test/wp-json/wc/v3/products/{product_id}?consumer_key=ck_xxxxx&consumer_secret=csxxxxx
Content-Type: application/json

{
  "date_on_sale_from_gmt":  "",
  "meta_data": [
    {
      "key": "test_key",
      "value": { "type":  "complex" }
    }
  ]
}
```
### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - Restore backward compatibility with WC 4.x and forward compatibility with WC 5.5

### CODE REVIEW

**Comment by vedanshujain** [comment]
Note that once this PR is approved, a corresponding PR in REST API repo will be created to target 4.3.3

**Review by peterfabian — CHANGES_REQUESTED** [review]
Do we need to also update this one? 
https://github.com/woocommerce/woocommerce/blob/master/includes/rest-api/Controllers/Version2/class-wc-rest-product-variations-v2-controller.php#L985

File: includes/rest-api/Controllers/Version2/class-wc-rest-product-variations-v2-controller.php [hunk 1]
```diff
@@ -981,7 +981,7 @@ public function get_item_schema() {
 								'context'     => array( 'view', 'edit' ),
 							),
 							'value' => array(
-								'description' => __( 'Meta value.', 'woocommerce' ),
+								'description' => __( 'Meta value.', 'woocommerce-rest-api' ),
```
Comment (peterfabian) [inline includes/rest-api/Controllers/Version2/class-wc-rest-product-variations-v2-controller.php:984]: I think we should keep this set to `woocommerce` domain, no?

File: includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php [hunk 2]
```diff
@@ -511,7 +511,7 @@ public function get_item_schema() {
 							),
 							'value' => array(
 								'description' => __( 'Meta value.', 'woocommerce' ),
-								'type'        => array( 'string', 'null' ),
+								'type'        => 'mixed',
```
Comment (peterfabian) [inline includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php:514]: I wonder if it would be worth adding a comment pointing to `adjust_wp_5_5_datatype_compatibility` or just mention why we're using `mixed` type even though it's not compatible with WP 5.5... but perhaps git blame works as well.

File: includes/rest-api/Controllers/Version3/class-wc-rest-controller.php [hunk 3]
```diff
@@ -93,16 +93,53 @@ public function get_endpoint_args_for_item_schema( $method = WP_REST_Server::CRE
...
 		}
 
+		$endpoint_args = $this->adjust_wp_5_5_datatype_compatibility( $endpoint_args );
+
+		return $endpoint_args;
+	}
+
+	/**
+	 * Change datatypes `date-time` to string, and `mixed` to composite of all built in types.
+	 *
```
Comment (peterfabian) [inline includes/rest-api/Controllers/Version3/class-wc-rest-controller.php:103]: Worth mentioning why we're doing this dynamically (the conflict between WP 5.4 and WP 5.5)?

**Comment by vedanshujain** [comment]
@peterfabian yup, good catch, looks like this was missed because of merged conflict. added now in https://github.com/woocommerce/woocommerce/pull/27318/commits/a4691dc3f375ef0135cf572200e5c9de1718101a (and test in https://github.com/woocommerce/woocommerce/pull/27318/commits/4294e06f7a8350dc7c47a60b64319440e1c5cd37)

File: tests/legacy/unit-tests/rest-api/Tests/Version3/products.php [hunk 4]
```diff
@@ -218,7 +219,7 @@ public function test_update_product() {
 		$product      = \Automattic\WooCommerce\RestApi\UnitTests\Helpers\ProductHelper::create_simple_product();
 		$response     = $this->server->dispatch( new WP_REST_Request( 'GET', '/wc/v3/products/' . $product->get_id() ) );
 		$data         = $response->get_data();
-		$date_created = date( 'Y-m-d\TH:i:s', current_time( 'timestamp' ) );
+		$date_created = gmdate( 'Y-m-d\TH:i:s', current_time( 'timestamp' ) );
```
Comment (peterfabian) [inline tests/legacy/unit-tests/rest-api/Tests/Version3/products.php:222]: Should we add a test for `null` and `""` just in case we forget about those again?

**Comment by Konamiman** [comment]
Verified that the example request gives a `400` in `master` but a `200` in this branch 👍🏻

**Review by Konamiman — APPROVED** [review]

File: tests/legacy/framework/traits/trait-wc-rest-api-complex-meta.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * Trait for easier testing of objects that have `mixed` data type somewhere.
+ *
+ * @package Automattic/WooCommerce/Tests/WC_REST_API_Complex_Meta.
+ */
+
+/**
+ * Trait WC_REST_API_Complex_Meta
+ */
+trait WC_REST_API_Complex_Meta {
```
Comment (peterfabian) [inline tests/legacy/framework/traits/trait-wc-rest-api-complex-meta.php:11]: Nice to trait it like this!

**Review by peterfabian — COMMENTED** [review]
Great job, all the test cases I've tried worked 100%! Just some suggestions for comments to provide more complex, but no stoppers, except for the domain I think.

File: tests/legacy/unit-tests/rest-api/Tests/Version3/products.php (hunk 4 above)
Comment (vedanshujain) [reply tests/legacy/unit-tests/rest-api/Tests/Version3/products.php:222]: agree, added in https://github.com/woocommerce/woocommerce/pull/27318/commits/dd9b9cfae2a3f565dddc90d135486c9cf1e58ba1

File: includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php (hunk 2 above)
Comment (vedanshujain) [reply includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php:514]: Agree, added more explanation in https://github.com/woocommerce/woocommerce/pull/27318/commits/3dd508fd356e44c8799ea7b29227ba04e6df2729.

File: includes/rest-api/Controllers/Version3/class-wc-rest-controller.php (hunk 3 above)
Comment (vedanshujain) [reply includes/rest-api/Controllers/Version3/class-wc-rest-controller.php:107]: Added further explanations in comments in https://github.com/woocommerce/woocommerce/pull/27318/commits/3dd508fd356e44c8799ea7b29227ba04e6df2729

File: includes/rest-api/Controllers/Version2/class-wc-rest-product-variations-v2-controller.php (hunk 1 above)
Comment (vedanshujain) [reply includes/rest-api/Controllers/Version2/class-wc-rest-product-variations-v2-controller.php]: good catch! fixed

**Review by peterfabian — APPROVED** [review]
Looks good now, thanks for the updates!
